### PR TITLE
support desc push_down

### PR DESCRIPTION
**Pull request by hanfei19910905**

### CODE REVIEW

File: store/localstore/local_client.go [hunk 1]
```diff
@@ -40,11 +43,20 @@ func (c *dbClient) SupportRequestType(reqType, subType int64) bool {
...
+			return true
+		}
 	}
 	return false
 }
 
+func (c *dbClient) SetSupportDesc(flag bool) {
+	SupportDesc = flag
+}
+
```
Comment (coocood) [inline store/localstore/local_client.go]: We do not need to set SupportDesc as a variable.

File: store/localstore/local_region.go [hunk 2]
```diff
@@ -329,7 +333,7 @@ func (rs *localRegion) getRowsFromIndexReq(txn kv.Transaction, sel *tipb.SelectR
 	return rows, nil
 }
 
-func getIndexRowFromRange(idxInfo *tipb.IndexInfo, txn kv.Transaction, ran kv.KeyRange) ([]*tipb.Row, error) {
+func getIndexRowFromRange(idxInfo *tipb.IndexInfo, txn kv.Transaction, ran kv.KeyRange, Desc bool) ([]*tipb.Row, error) {
```
Comment (coocood) [inline store/localstore/local_region.go]: Desc -> desc

File: executor/executor_xapi.go [hunk 3]
```diff
@@ -287,6 +288,8 @@ func (e *XSelectIndexExec) doIndexRequest() (*xapi.SelectResult, error) {
 	selIdxReq.StartTs = &startTs
 	selIdxReq.IndexInfo = tablecodec.IndexToProto(e.table.Meta(), e.indexPlan.Index)
 	selIdxReq.Limit = e.indexPlan.LimitCount
+	selIdxReq.OrderBy = make([]*tipb.ByItem, 0)
```
Comment (coocood) [inline executor/executor_xapi.go]: This line can be removed.

File: store/localstore/local_client.go [hunk 4]
```diff
@@ -171,6 +183,12 @@ func buildRegionTasks(client *dbClient, req *kv.Request) (tasks []*task) {
 			infoCursor++
 		}
 	}
+	if req.Desc && SupportDesc {
+		for i := 0; i < len(tasks); i++ {
```
Comment (coocood) [inline store/localstore/local_client.go]: len(tasks)/2

File: store/localstore/local_region.go [hunk 5]
```diff
@@ -319,8 +319,12 @@ func (rs *localRegion) getRowsFromIndexReq(txn kv.Transaction, sel *tipb.SelectR
 	idxID := sel.IndexInfo.GetIndexId()
 	kvRanges := rs.extractKVRanges(tid, idxID, sel.Ranges)
 	var rows []*tipb.Row
+	desc := false
+	if sel.OrderBy != nil {
+		desc = *sel.OrderBy[0].Desc
+	}
 	for _, ran := range kvRanges {
```
Comment (coocood) [inline store/localstore/local_region.go]: If `desc`, we should iterate kvRanges from last to first.

**Comment by coocood** [comment]
LGTM
@shenli PTAL

File: store/localstore/local_client.go [hunk 6]
```diff
@@ -12,6 +12,9 @@ type dbClient struct {
 	regionInfo []*regionInfo
 }
 
+// SupportDesc represents whether kv engine support read index reversely.
+var SupportDesc = true
```
Comment (shenli) [inline store/localstore/local_client.go]: Why not use SupportRequestType()?

Comment (hanfei19910905) [inline store/localstore/local_client.go]: because test will use it

**Comment by shenli** [comment]
LGTM

**Pull request merged**